File: Algorithm-Swift/Algorithm-Swift/Inflearn/60_합이같은부분집합.swift
//
//  60_합이같은부분집합.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/29.
//

import Foundation

var flag60 = false

func dfs60(l: Int, n: Int, arr: [Int], ch: inout [Int]) {
    if l == n {
        var sum1 = 0, sum2 = 0
        ch.enumerated().forEach { index, num in
            if num == 0 { sum1 += arr[index] }
            else { sum2 += arr[index] }
        }
        if sum1 == sum2 { flag60 = true }
        return
    } else {
        ch[l] = 1
        dfs60(l: l + 1, n: n, arr: arr, ch: &ch)
        ch[l] = 0
        dfs60(l: l + 1, n: n, arr: arr, ch: &ch)
    }
}

func solution60() {
    let n = Int(readLine()!)!
    let arr = readLine()!.split(separator: " ").map { Int($0)! }
    var ch = Array(repeating: 0, count: n)
    dfs60(l: 0, n: n, arr: arr, ch: &ch)

    if flag60 { print("YES") }
    else { print("NO") }
}

File: Algorithm-CPP/60_합이같은부분집합.cpp
#include <iostream>
using namespace std;

bool flag = false;

void dfs(int l, int n, int arr[], int ch[]) {
    if(l == n) {
        int sum1 = 0, sum2 = 0;
        for(int i = 0; i < n; i++) {
            if(ch[i] == 0) sum1 += arr[i];
            else sum2 += arr[i];
        }
        if(sum1 == sum2) flag = true;
        return;
    } else {
        ch[l] = 1;
        dfs(l + 1, n, arr, ch);
        ch[l] = 0;
        dfs(l + 1, n, arr, ch);
    }
}

int main() {
    int n;
    cin >> n;
    int arr[n], ch[n];
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
        ch[i] = 0;
    }
    
    dfs(0, n, arr, ch);

    if(flag) cout << "YES" << endl;
    else cout << "NO" << endl;

    return 0;
}